src/aspire/abinitio/commonline_d2.py

```diff
@@ -0,0 +1,1888 @@
+            Rik_J = J_conjugate(Rik)
+
+            final_votes = np.zeros(4)
+            final_votes[0] = self._compare_rots(Rij, Rjk_t, Rik)
+            final_votes[1] = self._compare_rots(Rij_J, Rjk_t, Rik)
+            final_votes[2] = self._compare_rots(Rij, Rjk_t_J, Rik)
+            final_votes[3] = self._compare_rots(Rij, Rjk_t, Rik_J)
+
+            J_list[trip_idx] = np.argmin(final_votes)
+            trip_idx += 1
+
+            pbar.update()
+        pbar.close()
+
+        return J_list
+
+    def _compare_rots(self, Rij, Rjk_t, Rik):
+        """
+        Compute norms for the 4 J-configurations and return indices
+        corresponding to best configuration for the provided triplet
+        of relative rotations.
+        """
+        # We compute the four sets of 4^3 norms |Rik @ Rjk.T - Rij|
+        # See equation (6.11) in publication.
+        prod_arr = Rik[:, None] @ Rjk_t[None]
+
+        arr = np.zeros((8, 8, 3, 3), dtype=self.dtype)
+        arr[0:4, 0:4] = prod_arr - Rij[0]
+        arr[0:4, 4:8] = prod_arr - Rij[1]
+        arr[4:8, 0:4] = prod_arr - Rij[2]
+        arr[4:8, 4:8] = prod_arr - Rij[3]
+
+        arr = arr.reshape((64, 9))
+        arr = np.sum(arr**2, axis=1)
+
+        m = np.sort(arr.flatten())
+        vote = np.sum(m[:16])
+
+        return vote
+
+    def _J_sync_power_method(self, J_list):
+        """
+        Calculate the leading eigenvector of the J-synchronization matrix
+        using the power method.
+
+        As the J-synchronization matrix is of size (n-choose-2)x(n-choose-2), we
+        use the power method to compute the eigenvalues and eigenvectors,
+        while constructing the matrix on-the-fly.
+
+        :param Rijs: (n-choose-2)x3x3 array of estimates of relative orientation matrices.
+
+        :return: An array of length n-choose-2 consisting of 1 or -1, where the sign
+            of the i'th entry indicates whether the i'th relative orientation matrix
+            will be J-conjugated.
+        """
+
+        # Set power method tolerance and maximum iterations.
+        epsilon = self.epsilon
+        max_iters = 100
+
+        # Initialize candidate eigenvectors
+        vec = randn(self.n_pairs, seed=self.seed)
+        vec = vec / norm(vec)
+        residual = 1
+        itr = 0
+
+        # Power method iterations
+        logger.info(
+            "Initiating power method to estimate J-synchronization matrix eigenvector."
+        )
+        while itr < max_iters and residual > epsilon:
+            itr += 1
+            vec_new = self._signs_times_v2(J_list, vec)
+            vec_new = vec_new / norm(vec_new)
+            residual = norm(vec_new - vec)
+            vec = vec_new
+            logger.info(
+                f"Iteration {itr}, residual {round(residual, 5)} (target {epsilon})"
+            )
+
+        # We need only the signs of the eigenvector
+        J_sync = np.sign(vec)
+        J_sync = np.sign(J_sync[0]) * J_sync  # Stabilize J_sync
```

Comment:
This is just to choose the eigenvector with positive first entry. It is not necessary, but I added it to get consistent results for the set of rotations that were being conjugated. Can remove, but will wait until after other changes.

Reply:
What happens if its negative?

Reply:
The J-sync method partitions the set of rotations into two sets and J-conjugates one of them. If we flipped the eigenvector to start with a negative entry, the other set of rotations would get J-conjugated. This wouldn't be a problem since the choice of which set to conjugate is arbitrary.

Reply:
Got it. If it helps give more determinism but has not other ill effects I'm in favor of leaving it.  Maybe we should add to other CL methods as we go.  Thanks!